apps/src/javalab/JavabuilderConnection.js

```diff
@@ -37,32 +41,55 @@ export default class JavabuilderConnection {
     } else {
       url += `?Authorization=${token}`;
     }
+
     this.socket = new WebSocket(url);
+    this.socket.onopen = this.onOpen.bind(this);
+    this.socket.onmessage = this.onMessage.bind(this);
+    this.socket.onclose = this.onClose.bind(this);
+    this.socket.onerror = this.onError.bind(this);
+  }
 
-    this.socket.onopen = () => {
-      this.onOutputMessage('Compiling...');
-    };
+  onOpen() {
+    this.onOutputMessage('Compiling...');
+  }
 
-    this.socket.onmessage = event => {
-      this.onOutputMessage(event.data);
-    };
+  onMessage(event) {
+    const data = JSON.parse(event.data);
+    switch (data.type) {
+      case WebSocketMessageType.SYSTEM_OUT:
+        this.onOutputMessage(data.value);
+        break;
+      case WebSocketMessageType.NEIGHBORHOOD:
+        Neighborhood.handleSignal(data);
+        break;
+      case WebSocketMessageType.EXCEPTION:
+        handleException(data, this.onOutputMessage);
+        break;
+      case WebSocketMessageType.DEBUBG:
+        this.onOutputMessage('--- Localhost debugging message ---');
```

Comment:
will students ever see a debug message or is this only for us?

Reply:
This is only for us. I'll add a check to make sure we never accidentally print this.